pkg/controller/test-images-distributor/test_images_distributor.go

```diff
@@ -0,0 +1,364 @@
+	return nil
+}
+
+func (r *reconciler) isImageStreamTagCurrent(
+	name types.NamespacedName,
+	targetClient ctrlruntimeclient.Client,
+	reference *imagev1.ImageStreamTag,
+) (bool, error) {
+
+	imageStreamTag := &imagev1.ImageStreamTag{}
+	if err := targetClient.Get(r.ctx, name, imageStreamTag); err != nil {
+		if apierrors.IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("failed to get imagestreamtag %s: %w", name.String(), err)
+	}
+
+	return imageStreamTag.Image.Name == reference.Image.Name, nil
+}
+
+const pullSecretName = "registry-cluster-pull-secret"
+
+func (r *reconciler) ensureImagePullSecret(namespace string, client ctrlruntimeclient.Client) error {
+	referenceSecret := r.generateReferencPullSecret(namespace)
+	name := types.NamespacedName{Namespace: namespace, Name: pullSecretName}
+	secret := &corev1.Secret{}
+	if err := client.Get(r.ctx, name, secret); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return fmt.Errorf("failed to get secret %s: %w", name.String(), err)
+		}
+		// Tolerate IsAlreadyExist, another routine might have created it or our cache is stale
+		if err := client.Create(r.ctx, referenceSecret); err != nil && !apierrors.IsAlreadyExists(err) {
+			return fmt.Errorf("failed to create secret %s/%s: %v", referenceSecret.Namespace, referenceSecret.Name, err)
+		}
+	}
+
+	// Make it comparable by resetting the ObjectMeta fields the apiserver sets
+	resourceVersion := secret.ObjectMeta.ResourceVersion
+	secret.ObjectMeta = referenceSecret.ObjectMeta
+	if !apiequality.Semantic.DeepEqual(secret, referenceSecret) {
+		referenceSecret.ResourceVersion = resourceVersion
+		if err := client.Update(r.ctx, referenceSecret); err != nil {
+			return fmt.Errorf("failed to update secret %s/%s: %w", referenceSecret.Namespace, referenceSecret.Name, err)
+		}
+	}
+
+	return nil
+}
+
+func (r *reconciler) generateReferencPullSecret(namespace string) *corev1.Secret {
+	return &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      pullSecretName,
+		},
+		Data: map[string][]byte{
+			corev1.DockerConfigJsonKey: r.pullSecretGetter(),
+		},
+		Type: corev1.SecretTypeDockerConfigJson,
+	}
+}
+
+func testInputImageStreamTagFilterFactory(l *logrus.Entry, ca agents.ConfigAgent) (objectFilter, error) {
+	const indexName = "config-by-test-input-imagestreamtag"
+	if err := ca.AddIndex(indexName, indexConfigsByTestInputImageStramTag); err != nil {
+		return nil, fmt.Errorf("failed to add %s index to configAgent: %w", indexName, err)
+	}
+	l = logrus.WithField("subcomponent", "test-input-image-stream-tag-filter")
+	return func(nn types.NamespacedName) bool {
+		result, err := ca.GetFromIndex(indexName, nn.String())
+		// Today, GetFromIndex only errors if the index does not exist, so this should
+		// never happen.
+		if err != nil {
+			l.WithField("name", nn.String()).WithError(err).Error("Failed to get configs from index")
+			return false
+		}
+		return len(result) > 0
```

Comment:
This is not extremely smart or efficient, but the best we can do right now without having to extend the config agent with support for generic post-config-load hooks. IMHO its good enough and I'd prefer to not extend the public api of the config agent.